### ISSUE TITLE
`class extends null` is unsupported by ECMAScript

### ISSUE THREAD

**Issue by shinenelson** [tddbin/katas#98]
`class extends null` was reverted by ECMAScript in https://github.com/tc39/ecma262/pull/781 ( https://github.com/tc39/ecma262/commit/c57ef95c45a371f9c9485bb1c3881dbdc04524a2 ) and is [still broken](https://github.com/tc39/ecma262/issues/1036).

The [kata](https://github.com/tddbin/katas/blob/master/katas/es6/language/class/extends.js#L19-L25) is abnormal since it doesn't have a solution in the test itself.
So, the question is whether we should have that test for `class extends null`?

However, I found a possible solution thanks to [devdocs.io](https://devdocs.io/javascript/classes/extends#Extending_null) ( which had a previous version of the [MDN Web Docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/extends); [revision](https://wiki.developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/extends$compare?locale=en-US&to=1621409&from=1611436) ) : 
```js
class NullClass extends null {}
assert.equal(Object.getPrototypeOf(NullClass), Function.prototype);
```

**Comment by wolframkriesing** [comment]
if i read the spec (I am not sure if I read the wrong part) it looks like its is possible
https://tc39.es/ecma262/#sec-runtime-semantics-classdefinitionevaluation
in 5e it says "If superclass is null, then" and [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/extends) also still says "The .prototype of the extension must be an Object or null."

Do I read those wrong?

**Comment by wolframkriesing** [comment]
though i dont get it work anywhere neither
```
> class A extends null {}
undefined
> new A
VM79:1 Uncaught TypeError: Super constructor null of A is not a constructor
    at new A (<anonymous>:1:1)
    at <anonymous>:1:1
```

**Comment by wolframkriesing** [comment]
yeah the only useful thing to learn is what you described. But not really of any practical value, is it?

So I would rather change this test to verify that a parent must be callable with `super()` and since `null` has no super it throws. Does that makes sense?

**Comment by shinenelson** [comment]
I'm not really sure. Wouldn't that be kind of misleading new developers to assume that a `class` _may_ extend `null`?

We should either remove the whole test or explain the rationale behind the use-case. And also somehow probably document somewhere ( in the test itself ) that this is a _gray-area_ test.

> Do I read those wrong?

No, what you're reading is correct. The spec was first made to extend null, but then apparently many bad things happened and they reverted it ( see links in issue description )

I found [this question on StackOverflow](https://stackoverflow.com/questions/41189190/how-and-why-would-i-write-a-class-that-extends-null) that questions this use-case altogether. The question has different errors of this use-case _not_ working

PS : the issue title is probably misleading, what I meant was that it was broken by ECMAScript and that it wasn't our fault.

**Comment by wolframkriesing** [comment]
i agree, either 1) remove the test or 2) make it more explicit, maybe like that:

```
it('`extends null` is not supported', () => {
      class NullClass extends null {}
      assert.throws(() => new NullClass);
});
```

just to mention the edge case and its solution.
What do you think?

**Comment by shinenelson** [comment]
actually, it is supported by the spec, it's just not working as intended. So, maybe, the test should be **`extends null` is broken** or something along those lines maybe?

**Comment by shinenelson** [comment]
the [MDN revision](https://wiki.developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/extends$compare?locale=en-US&to=1621409&from=1611436) used the term 'not allowed', so, I guess 'not supported' is close enough. I'd favor 'not allowed' though.
